pkg/objstore/azure/azure.go

```diff
@@ -0,0 +1,292 @@
+
+	level.Debug(logger).Log("msg", "Creating new Azure bucket connection", "component", component)
+
+	var conf Config
+	if err := yaml.Unmarshal(azureConfig, &conf); err != nil {
+		return nil, err
+	}
+
+	err := conf.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := context.Background()
+
+	container, err := createContainer(ctx, conf.StorageAccountName, conf.StorageAccountKey, conf.ContainerName)
+	if err != nil {
+		if err.(blob.StorageError).ServiceCode() == "ContainerAlreadyExists" {
+			level.Debug(logger).Log("msg", "Using existing Azure blob container", "address", container)
+			level.Debug(logger).Log("msg", "Getting connection to existing Azure blob container", "container", conf.ContainerName)
+			container, err = getContainer(ctx, conf.StorageAccountName, conf.StorageAccountKey, conf.ContainerName)
+			if err != nil {
+				return nil, errors.Wrapf(err, "msg", "Cannot get existing Azure blob container", "address", container)
+			}
+		} else {
+			return nil, errors.Wrapf(err, "msg", "Error creating Azure blob container", "address", container)
+		}
+	} else {
+		level.Info(logger).Log("msg", "Azure blob container successfully created", "address", container)
+	}
+
+	bkt := &Bucket{
+		logger:       logger,
+		containerURL: container,
+		config:       &conf,
+		opsTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name:        "thanos_objstore_azure_storage_operations_total",
+			Help:        "Total number of operations that were executed against an Azure storage account.",
+			ConstLabels: prometheus.Labels{"storage_account": conf.StorageAccountName},
+		}, []string{"operation"}),
+	}
+
+	if reg != nil {
+		reg.MustRegister(bkt.opsTotal)
+	}
+	return bkt, nil
+}
+
+// Iter calls f for each entry in the given directory. The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	b.opsTotal.WithLabelValues(opObjectsList).Inc()
+
+	prefix := dir
+	if prefix != "" && !strings.HasSuffix(prefix, DirDelim) {
+		prefix += DirDelim
+	}
+
+	list, err := b.containerURL.ListBlobsHierarchySegment(ctx, blob.Marker{}, DirDelim, blob.ListBlobsSegmentOptions{
+		Prefix: prefix,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "msg", "Cannot list blobs in directory %s", dir)
+	}
+	var listNames []string
+
+	for _, blob := range list.Segment.BlobItems {
+		listNames = append(listNames, blob.Name)
+	}
+
+	for _, blobPrefix := range list.Segment.BlobPrefixes {
+		listNames = append(listNames, blobPrefix.Name)
+	}
+
+	for _, name := range listNames {
+		if err := f(name); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// IsObjNotFoundErr returns true if error means that object is not found. Relevant to Get operations.
+func (b *Bucket) IsObjNotFoundErr(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	errorCode := parseError(err.Error())
```

Comment:
This is not what I commented about (: The solution is just to compare errors variables directly. See this https://github.com/improbable-eng/thanos/blob/master/pkg/objstore/gcs/gcs.go#L174

Reply:
Honestly, I spent quite a lot of time dealing with Azure API errors. The thing is there is no way to do that in Azure SDK, at least for now. There is no errors library or model classes :( It also returns proper request object only if request is succeeded otherwise request is nil, so you can't grab the status code from there :(  I know it's looks hacky but that's where I end up after few weeks of try and fail.

Reply:
I guess because Golang Azure SDK is autogenerated library. :(

Reply:
Ah, I thought we are in control of the errors... ok let's stick to that then.